compositor_common/src/validators.rs

```diff
@@ -98,4 +115,269 @@ impl SceneSpec {
 
         Ok(())
     }
+
+    fn validate_cycles(
+        &self,
+        registered_inputs: &HashSet<NodeId>,
+        transform_nodes: &HashMap<NodeId, &TransformNodeSpec>,
+    ) -> Result<(), SpecValidationError> {
+        enum NodeState {
+            BeingVisited,
+            Visited,
+        }
+
+        let mut visited: HashMap<NodeId, NodeState> = HashMap::new();
+
+        fn visit(
+            node: &NodeId,
+            transform_nodes: &HashMap<NodeId, &TransformNodeSpec>,
+            inputs: &HashSet<NodeId>,
+            visited: &mut HashMap<NodeId, NodeState>,
+        ) -> Result<(), SpecValidationError> {
+            if let Some(_input) = inputs.get(node) {
+                return Ok(());
+            }
+
+            match visited.get(node) {
+                Some(NodeState::BeingVisited) => return Err(SpecValidationError::CycleDetected),
+                Some(NodeState::Visited) => return Ok(()),
+                None => {}
+            }
+
+            visited.insert(node.clone(), NodeState::BeingVisited);
+
+            for child in &transform_nodes.get(node).unwrap().input_pads {
+                visit(child, transform_nodes, inputs, visited)?;
+            }
+
+            visited.insert(node.clone(), NodeState::Visited);
+
+            Ok(())
+        }
+
+        for output in &self.outputs {
+            visit(
+                &output.input_pad,
+                transform_nodes,
+                registered_inputs,
+                &mut visited,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn validate_node_ids_uniqueness(
+        &self,
+        defined_node_ids: &Vec<NodeId>,
+    ) -> Result<(), SpecValidationError> {
+        let mut nodes_ids: HashSet<NodeId> = HashSet::new();
```

Comment:
```suggestion
        let mut nodes_ids: HashSet<&NodeId> = HashSet::new();
```
clone only when returning error, but there is no need for that to just keep track of which nodes were visited

Reply:
I modified those functions' lifetimes and removed unnecessary cloning